File: cpp/tests/l3_fixture.cpp
#include "llab/order_book.hpp"

#include <cstdint>
#include <fstream>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>

namespace {

using llab::order_book::Event;
using llab::order_book::EventType;
using llab::order_book::OrderBook;
using llab::order_book::Side;

Side parse_side(const std::string &text) {
    if (text == "B")
        return Side::Bid;
    if (text == "A")
        return Side::Ask;
    throw std::runtime_error("invalid side");
}

void require(bool condition, const char *message) {
    if (!condition)
        throw std::runtime_error(message);
}

} // namespace

int main(int argc, char **argv) {
    if (argc != 2)
        return 2;
    std::ifstream input(argv[1]);
    if (!input)
        return 2;

    OrderBook book;
    std::string line;
    bool header_seen = false;
    while (std::getline(input, line)) {
        if (line.empty() || line.starts_with('#'))
            continue;
        if (!header_seen) {
            require(line == "LLBT/1", "unexpected trace version");
            header_seen = true;
            continue;
        }
        if (line.starts_with("EXPECT ")) {
            std::uint64_t count, bid_price, bid_quantity, ask_price, ask_quantity, digest;
            std::istringstream expected(line.substr(7));
            expected >> count >> bid_price >> bid_quantity >> ask_price >> ask_quantity >> digest;
            require(book.live_order_count() == count, "live-order count mismatch");
            require(book.best_bid() ==
                        std::optional{llab::order_book::Level{bid_price, bid_quantity}},
                    "best bid mismatch");
            require(book.best_ask() ==
                        std::optional{llab::order_book::Level{ask_price, ask_quantity}},
                    "best ask mismatch");
            std::cout << book.state_digest() << '\n';
            if (digest != 0)
                require(book.state_digest() == digest, "state digest mismatch");
            return 0;
        }
        std::istringstream fields(line);
        char kind = '\0';
        fields >> kind;
        std::uint64_t id = 0, id2 = 0, price = 0, quantity = 0;
        std::string side;
        if (kind == 'A') {
            fields >> id >> side >> price >> quantity;
            book.apply({EventType::Add, id, 0, parse_side(side), price, quantity});
        } else if (kind == 'C' || kind == 'E') {
            fields >> id >> quantity;
            book.apply({kind == 'C' ? EventType::Cancel : EventType::Execute, id, 0, Side::Bid, 0,
                        quantity});
        } else if (kind == 'D' || kind == 'X') {
            fields >> id;
            book.apply({kind == 'D' ? EventType::Delete : EventType::OrderDelete, id});
        } else if (kind == 'R') {
            fields >> id >> id2 >> side >> price >> quantity;
            book.apply({EventType::Replace, id, id2, parse_side(side), price, quantity});
        } else if (kind == 'U') {
            fields >> id >> side >> price >> quantity;
            book.apply({EventType::OrderUpsert, id, 0, parse_side(side), price, quantity});
        } else {
            throw std::runtime_error("invalid trace record");
        }
    }
    return 2;
}

File: rust/tests/l3_fixture.rs
use low_latency_lab_benchmarks::order_book::{Event, Level, OrderBook, Side};

fn side(text: &str) -> Side {
    match text {
        "B" => Side::Bid,
        "A" => Side::Ask,
        _ => panic!("invalid side"),
    }
}

#[test]
fn l3_smoke_fixture_matches_contract() {
    let trace = include_str!("../../data/fixtures/l3_smoke.llbt");
    let mut book = OrderBook::default();
    let mut header_seen = false;

    for line in trace.lines() {
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        if !header_seen {
            assert_eq!(line, "LLBT/1");
            header_seen = true;
            continue;
        }
        let fields: Vec<_> = line.split_ascii_whitespace().collect();
        match fields[0] {
            "A" => book.apply(Event::Add {
                order_id: fields[1].parse().unwrap(),
                side: side(fields[2]),
                price_ticks: fields[3].parse().unwrap(),
                quantity: fields[4].parse().unwrap(),
            }),
            "C" => book.apply(Event::Cancel {
                order_id: fields[1].parse().unwrap(),
                quantity: fields[2].parse().unwrap(),
            }),
            "E" => book.apply(Event::Execute {
                order_id: fields[1].parse().unwrap(),
                quantity: fields[2].parse().unwrap(),
            }),
            "D" => book.apply(Event::Delete { order_id: fields[1].parse().unwrap() }),
            "R" => book.apply(Event::Replace {
                old_order_id: fields[1].parse().unwrap(),
                new_order_id: fields[2].parse().unwrap(),
                side: side(fields[3]),
                price_ticks: fields[4].parse().unwrap(),
                quantity: fields[5].parse().unwrap(),
            }),
            "U" => book.apply(Event::OrderUpsert {
                order_id: fields[1].parse().unwrap(),
                side: side(fields[2]),
                price_ticks: fields[3].parse().unwrap(),
                quantity: fields[4].parse().unwrap(),
            }),
            "X" => book.apply(Event::OrderDelete { order_id: fields[1].parse().unwrap() }),
            "EXPECT" => {
                assert_eq!(book.live_order_count(), fields[1].parse::<usize>().unwrap());
                assert_eq!(
                    book.best_bid(),
                    Some(Level {
                        price_ticks: fields[2].parse().unwrap(),
                        quantity: fields[3].parse().unwrap()
                    })
                );
                assert_eq!(
                    book.best_ask(),
                    Some(Level {
                        price_ticks: fields[4].parse().unwrap(),
                        quantity: fields[5].parse().unwrap()
                    })
                );
                let expected_digest: u64 = fields[6].parse().unwrap();
                if expected_digest != 0 {
                    assert_eq!(book.state_digest(), expected_digest);
                }
                return;
            }
            other => panic!("invalid trace record: {other}"),
        }
        .unwrap();
    }
    panic!("fixture has no EXPECT record");
}
